templates/load-message.lua

```diff
@@ -0,0 +1,57 @@
+-- Some parameters to control the behavior. Can be set via Redis keys.
+-- How long (in seconds) should a given buffer bucket be?
+local pBufferBucketInterval = tonumber(redis.call("GET", "conf-bufferBucketInterval")) or 60
+-- How many buffer buckets should we retain?
+local pBufferBucketCount = tonumber(redis.call("GET", "conf-bufferBucketCount")) or 20
+
+-- Derived parameters
+local pBufferLifetime = pBufferBucketInterval * pBufferBucketCount
+
+-- First, parse and restructure the incoming message.
+local message = cjson.decode(ARGV[1])
+message.redisMessageId = tonumber(redis.call("INCR", "last-redisMessageId"))
+local messageJson = cjson.encode(message)
+
+-- It's OK to index messages by app name, since app names are unique within an
+-- environment.
+local appName = message.app
+
+-- Now, make sure we deliver the message to subscribers waiting for it.
+local publishRet = redis.call("PUBLISH", "stream-" .. appName, messageJson)
+
+-- Now, we need to add the key to a buffer. We actually want to use multiple buffers
+-- for each log stream, so that the Redis volatile-ttl policy will automatically
+-- expire old logs for us, without entirely trimming the buffers. To that end, we
+-- buffer logs into a new bucket every minute, which means we'll have 10 active
+-- buckets per log stream, since our TTL on log buffers is 10 minutes.
+local bufferBucketIndex = math.floor(message.unix_timestamp / pBufferBucketInterval) * pBufferBucketInterval
+local bufferBucketName = "buffer-bucket-" .. appName .. "-" .. bufferBucketIndex
+
+-- New messages are added at the right end of the bucket, which means that
+-- messages will be sorted from first received to last received (note: the
+-- ordering within a bucket is based on the order in which the messages were
+-- received, not their actual timestamp).
+redis.call("RPUSH", bufferBucketName, messageJson)
+
+-- Set a  TTL, which ensures we eventually expire the bucket, and will be used
+-- by Redis to inform OOM eviction decisions (oldest buckets - which will
+-- expire the soonest - are evicted first).
+redis.call("EXPIREAT", bufferBucketName, message.unix_timestamp + pBufferLifetime)
+
+-- Finally, store the bucket name in a buffer map. The buffer map is a sorted
+-- set where we sort bucket names by their timestamp. Thus, when we query the
+-- set via ZRANGE ... 0 -1, we'll get the buckets from oldest to newest.
+local bufferMapName = "buffer-map-" .. appName
```

Comment:
Why do you need the buffer map? It looks like the buffer buckets all have appropriately staggered TTLs set, couldn't you use a `SCAN buffer-bucket-#{appName}-*` instead to query the current buckets in dumptruck and avoid maintaining the buffer map?